src/builder/modules/Generator.res

```diff
@@ -122,8 +122,49 @@ let usePage = id => {
   model := Builder.usePage(model.contents, id)
 }
 
-let fillBackgroundColor = (fillStyle: string) => {
-  model := Generator_Builder.fillBackgroundColor(model.contents, fillStyle)
+let fillBackgroundColor = (color: string) => {
+  model := Generator_Builder.fillBackgroundColor(model.contents, color)
+}
+
+let fillBackgroundColorWithWhite = () => {
+  model := Generator_Builder.fillBackgroundColor(model.contents, "#ffffff")
+}
+
+let drawLine = (
+  from: Builder.position,
+  to: Builder.position,
+  ~color: string="#000000",
+  ~width: float=1.0,
+  ~pattern: array<int>=[],
```

Comment:
Since in this PR we have a great standard function for drawing fold lines, so I am wondering if there is any need for the line drawing function to support arbitrary patterns?

I'm mindful of keeping the generator functions to things we know are needed (it's a good principle in software design in general to help simplify systems).

Is there a particular idea you have in mind for the pattern support? If no particular cases, then let's remove the `pattern` and `offset` options.

Btw, still good to keep those options on the `Generator_Builder` since that's internal and not exposed to the generator writers.

Reply:
There are two places I can think of where having custom patterns and offsets will come in handy:

- In the action figure generator, there is a unique dash pattern used to subtly indicate that a fold line needs to also be cut- the drawLine() functions should be able to draw these when needed.
- My idea for an outline drawing function will have to be able to draw paths with gaps- e.g. when there is a hole in an item texture that is drawn on, or when drawing along a higher-resolution diagonal edge. In these cases, the line will need to use the offset to make it look like it is continued from the previous line instead of just starting over.

Reply:
> In the action figure generator, there is a unique dash pattern used to subtly indicate that a fold line needs to also be cut- the drawLine() functions should be able to draw these when needed.

That's interesting. I wonder if we need the idea of a `cut` line then? I know it's not used a lot, but it would be good to standardise that meaning in all generators. How about we provide `drawFoldLine()` and `drawCutLine()` functions.

Btw, I'm thinking we rename "drawFold" to "drawFoldLine" - it's a bit more accurate to what it is, and fits will with the naming of the other line function(s).

> My idea for an outline drawing function will have to be able to draw paths with gaps- e.g. when there is a hole in an item texture that is drawn on

Regarding the gaps, couldn't that be done with separate fold lines? I suspect that will be easier than trying to work out pattern offset values? Btw, I'm understanding the offsets with a gap might look something like this `[2, 2, 2, 2, 2, 2, 2, 20, 2, 2, 2, 2, 2]` but I might not have the idea right?

> or when drawing along a higher-resolution diagonal edge. In these cases, the line will need to use the offset to make it look like it is continued from the previous line instead of just starting over.

I am not quite following the idea here. Maybe an image example will help? Could yo describe how the regular fold line might not be suitable in these cases? I am also not too concerned about pixel perfection, ultimately this is printed and then cut with scissors/knife which are not precision devices, but maybe I am missing something here.

Reply:
> That's interesting. I wonder if we need the idea of a cut line then? I know it's not used a lot, but it would be good to standardise that meaning in all generators. How about we provide drawFoldLine() and drawCutLine() functions.

Cut lines aren't used all that often in designs, and when they are they are usually represented by just a red line- The subtle fold-and-cut line is more of a specific situation that should be where the pattern and offset parameters are used.

> Btw, I'm thinking we rename "drawFold" to "drawFoldLine" - it's a bit more accurate to what it is, and fits will with the naming of the other line function(s).

This does make sense, especially considering derivative functions where it would make more sense to have "drawRect" and "drawFoldRect" than "drawLineRect" and "drawFoldRect", etc.

> I am not quite following the idea here. Maybe an image example will help? Could yo describe how the regular fold line might not be suitable in these cases? I am also not too concerned about pixel perfection, ultimately this is printed and then cut with scissors/knife which are not precision devices, but maybe I am missing something here.

Take for example, a higher resolution texture that might need an outline drawn around it: 
![diamond](https://user-images.githubusercontent.com/16890123/170856575-8f48d266-84c2-450a-80cd-cd539cc88e3b.png)

Though the line functions can draw diagonal lines, the outline function will probably generate and follow a path composed of points that are all at right angles from each other, meaning that slopes may draw new lines at every row they are on. With a default unchanging offset, the fold line on an edge of this texture may look like this: 
![diamond-nooffset](https://user-images.githubusercontent.com/16890123/170856854-637a89ed-df20-419b-a339-a2cc85051773.png)

If the offsets are changed to be related to where along the path the line starts, it may look like this: 
![diamond-yesoffset](https://user-images.githubusercontent.com/16890123/170856981-02fe1d8a-fcc8-4374-bab8-9b4ae2ca1e78.png)

This way, the 2 - 2 pattern still repeats although the line technically is not continuous and is actually restarting at every new y value.

Reply:
> Cut lines aren't used all that often in designs, and when they are they are usually represented by just a red line- The subtle fold-and-cut line is more of a specific situation that should be where the pattern and offset parameters are used.

No worries, even though they are not very common it would be good to standardise them. This makes it trivial for newcomers plus keeps a convention through all generators. As new requirements come up we can add more functions, but I'm not sure there is a genuine need to have arbitrary patterns? I am happy for you to choose what those standard fold/cut patterns might be (and the function names) - presumably using the same colour as you already have here?

> Take for example, a higher resolution texture that might need an outline drawn around it:

Thanks for taking the time to create those example images and explain the problem in detail. I suspect we won't use the existing drawFold() or drawLine() function for doing outlines. It's more likely this will take some different strategy inside the generator itself, so let's not influence the drawLine just yet until we have a better understanding of the outline coding strategy.

Unless there is a compelling reason to keep the pattern, then let's leave them out for now. We can always revisit later if it comes up again.